File: zig/src/input.zig
const chip8 = @import("chip8.zig");
const sdl = @import("sdl.zig");
const std = @import("std");

pub const InputError = error{
    UnknownAction,
};

pub fn handleKey(cpu: *chip8.Cpu, action: sdl.SDL_EventType, key: sdl.SDL_Keycode) InputError!void {
    var keyValue: u8 = undefined;
    if (action == sdl.SDL_KEYDOWN) {
        keyValue = 1;
    } else if (action == sdl.SDL_KEYUP) {
        keyValue = 0;
    } else {
        return InputError.UnknownAction;
    }

    const keycode: u8 = switch (key) {
        sdl.SDLK_1 => 0x1,
        sdl.SDLK_2 => 0x2,
        sdl.SDLK_3 => 0x3,
        sdl.SDLK_4 => 0xC,
        sdl.SDLK_q => 0x4,
        sdl.SDLK_w => 0x5,
        sdl.SDLK_e => 0x6,
        sdl.SDLK_r => 0xD,
        sdl.SDLK_a => 0x7,
        sdl.SDLK_s => 0x8,
        sdl.SDLK_d => 0x9,
        sdl.SDLK_f => 0xE,
        sdl.SDLK_z => 0xA,
        sdl.SDLK_x => 0x0,
        sdl.SDLK_c => 0xB,
        sdl.SDLK_v => 0xF,
        else => return,
    };
    cpu.key[keycode] = keyValue;
}

test "handleKey" {
    var cpu = chip8.Cpu{};
    chip8.initialize(&cpu);
    try std.testing.expectError(InputError.UnknownAction, handleKey(&cpu, 99, sdl.SDLK_1));
    try std.testing.expectEqual(@as(u8, 0), cpu.key[0x1]);

    try handleKey(&cpu, sdl.SDL_KEYDOWN, sdl.SDLK_1);
    try std.testing.expectEqual(@as(u8, 1), cpu.key[0x1]);

    try handleKey(&cpu, sdl.SDL_KEYUP, sdl.SDLK_1);
    try std.testing.expectEqual(@as(u8, 0), cpu.key[0x1]);
}

File: c/src/input.c
#include "input.h"

struct input* input_init(void) {
    struct input* input = malloc(sizeof(struct input));
    if (!input) {
        return NULL;
    }
    *input = (struct input){.key = {0}};
    return input;
}

void input_handle_event(struct input* input, SDL_Event event) {
    bool key_value = false;
    if (event.type == SDL_KEYDOWN) {
        key_value = true;
    } else if (event.type != SDL_KEYUP) {
        printf("Unknown input event: %d", event.type);
        return;
    }

    uint8_t keycode = 0;
    switch (event.key.keysym.sym) {
    case SDLK_1:
        keycode = 0x1;
        break;
    case SDLK_2:
        keycode = 0x2;
        break;
    case SDLK_3:
        keycode = 0x3;
        break;
    case SDLK_4:
        keycode = 0xC;
        break;
    case SDLK_q:
        keycode = 0x4;
        break;
    case SDLK_w:
        keycode = 0x5;
        break;
    case SDLK_e:
        keycode = 0x6;
        break;
    case SDLK_r:
        keycode = 0xD;
        break;
    case SDLK_a:
        keycode = 0x7;
        break;
    case SDLK_s:
        keycode = 0x8;
        break;
    case SDLK_d:
        keycode = 0x9;
        break;
    case SDLK_f:
        keycode = 0xE;
        break;
    case SDLK_z:
        keycode = 0xA;
        break;
    case SDLK_x:
        keycode = 0x0;
        break;
    case SDLK_c:
        keycode = 0xB;
        break;
    case SDLK_v:
        keycode = 0xF;
        break;
    default:
        return;
    }
    input->key[keycode] = key_value;
}

void input_set_key(struct input* input, uint8_t keycode, bool key_value) {
    input->key[keycode] = key_value;
}

bool input_get_key(struct input* input, uint8_t keycode) {
    return input->key[keycode];
}
